### PR TITLE
Register with upload

### PR DESCRIPTION
**Pull request by janstaelensskyline**
Allows for Registration during the Artifact Uploading.
No changes made to the action arguments list.
No changes needed to WorkFlows.
No extra secrets and setup needed.

Docker updated to install GIT.
Uses GIT commands to retrieve Branch Name and Committer information.
Calls an updated ArtifactUploader API in Azure with new arguments to perform a clean registration to sharepoint tables:
CatalogMapping, Catalogs, CatalogVersions.

Note that this ArtifactUploader is (at this time) not released, but it is running on an environment 'test1'
To test this before the ArtifactUploader is released, you can change your workflow to
  CI:
    uses: SkylineCommunications/_ReusableWorkflows/.github/workflows/Automation Master Workflow.yml@RegisterOnUpload

the RegisterOnUpload has changes that remove most of the CI (for speed) and it will run this deploy action using the test environment. (on full release, the reusable branch can be deleted)

### CODE REVIEW

File: GitHubAction/GitHubAction/Factories/CatalogDataFactory.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            bool isPreRelease = CleanAndAddVersion(inputs, catalog);
+            catalog.IsPreRelease = isPreRelease;
+
+            return catalog;
+        }
+
+        private static bool CleanAndAddVersion(Inputs inputs, CatalogData catalog)
+        {
+            bool isPreRelease;
+            if (!string.IsNullOrWhiteSpace(inputs.BuildNumber))
```
Comment (PedroDebevere) [inline GitHubAction/GitHubAction/Factories/CatalogDataFactory.cs:44]: build number will only be filled in for prereleases?

File: GitHubAction/Catalog.Domain/CatalogData.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+                   Branch == data.Branch &&
+                   IsPreRelease == data.IsPreRelease &&
+                   Identifier == data.Identifier &&
+                   Name == data.Name &&
+                   CommitterMail == data.CommitterMail &&
+                   ReleaseUri == data.ReleaseUri &&
+                   ContentType == data.ContentType;
+        }
+
+        public override int GetHashCode()
```
Comment (PedroDebevere) [inline GitHubAction/Catalog.Domain/CatalogData.cs:27]: overriding GetHashCode based on members that can be modified should be avoided.

File: GitHubAction/GIT/GITInfo.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+                    powershell.AddScript("$PSVersionTable.PSVersion");
+                    var version = String.Join(",", powershell.Invoke());
+                    powershell.Commands.Clear();
+
+                    powershell.AddScript("$PSVersionTable.Platform");
+                    var checkOS = String.Join(",", powershell.Invoke());
+                    powershell.Commands.Clear();
+
+                    string resultString = "errors: " + String.Join(",", powershell.Streams.Error.ReadAll());
+                    throw new InvalidOperationException("GIT Initial Setup Failed with PowerShell Errors: " + resultString + Environment.NewLine + " ps version: " + version + " OS: " + checkOS + " git version:" + gitVersion + "--end");
```
Comment (PedroDebevere) [inline GitHubAction/GIT/GITInfo.cs:37]: use interpolated string?

Comment (janstaelensskyline) [reply GitHubAction/GIT/GITInfo.cs:37]: Yes. Still got 'old-school' string in my fingers when I don't pay attention.
Is the below ok? Not sure about Environment.NewLine

```suggestion
                    throw new InvalidOperationException($"GIT Initial Setup Failed with PowerShell Errors: {resultString} {Environment.NewLine} ps version: {version} OS: {checkOS} git version: {gitVersion} --end");
```

File: GitHubAction/GitHubAction/Factories/CatalogDataFactory.cs (hunk 1 above)
Comment (janstaelensskyline) [reply GitHubAction/GitHubAction/Factories/CatalogDataFactory.cs:44]: Build number is an optional argument in the github action. It's filled in when there is no version tag but only a branch and a build number.

File: GitHubAction/Catalog.Domain/CatalogData.cs (hunk 2 above)
Comment (PedroDebevere) [reply GitHubAction/Catalog.Domain/CatalogData.cs:27]: can you make the type immutable? then you can leave the gethashcode

Comment (janstaelensskyline) [reply GitHubAction/Catalog.Domain/CatalogData.cs]: Unable to make Immutable. The object gets filled in over time.
I've left the code as-is for now, but I've added a comment to indicate the CatalogData object should not be used as the Key of a dictionary or be placed in a hashset.

**Review by PedroDebevere — APPROVED** [review]

**Pull request merged**